RHIDP-1925 bulk import from GitHub (#616)
redhat-developer/red-hat-developers-documentation-rhdh: modules/importing-repositories/procedure-enabling-the-bulk-import-from-github-feature.adoc

[id="enabling-ang-giving-access-to-the-bulk-import-feature"]
= Enabling and giving access to the Bulk Import feature

.Procedure
. To give {product-short} access to GitHub, link:{linkgettingstartedguide}#configuring-github-integration[configure GitHub integration].

. The Bulk Import plugin is installed but disabled by default.

> Enable the `./dynamic-plugins/dist/janus-idp-backstage-plugin-bulk-import-backend-dynamic` and `./dynamic-plugins/dist/janus-idp-backstage-plugin-bulk-import` preinstalled plugins.

Author: @rm3l https://docs.redhat.com/en/documentation/red_hat_developer_hub/1.3/html-single/dynamic_plugins_reference/index#rhdh-tech-preview-plugins mentions only one plugin:

> Bulk Import
> @janus-idp/backstage-plugin-bulk-import
> 1.4.9
> ./dynamic-plugins/dist/janus-idp-backstage-plugin-bulk-import

Is the second plugin missing in the table, or has it been removed?

Reviewer: The backend plugin is currently in the list of [community plugins](https://docs.redhat.com/en/documentation/red_hat_developer_hub/1.3/html-single/dynamic_plugins_reference/index#rhdh-community-plugins).
This was a mistake that is already fixed in https://github.com/janus-idp/backstage-plugins/pull/2223 (but for 1.4 (where both plugins will have the same support level - Tech Preview)

> See link:{plugins-configure-book-url}[{plugins-configure-book-title}].

Author: There is no docs to explain how to enable this plugin specifically in https://docs.redhat.com/en/documentation/red_hat_developer_hub/1.3/html-single/configuring_dynamic_plugins/index

It might make sense to provide an app-config YAML fragment here. @rm3l Can you provide the required YAML fragment?

Reviewer: Yes, that's a good point. I thought we had a generic doc on how to enable a pre-installed plugin..

- So for Helm, that would mean adding this to the values file:
```yaml
global:
  dynamic:
    includes:
      - dynamic-plugins.default.yaml
    plugins:
      ### Bulk Import
      - package: ./dynamic-plugins/dist/janus-idp-backstage-plugin-bulk-import-backend-dynamic
        disabled: false
      - package: ./dynamic-plugins/dist/janus-idp-backstage-plugin-bulk-import
        disabled: false
# [...]
```

- And for the Operator, adding this to the dynamic-plugins ConfigMap (and making sure the ConfigMap is referenced in the `spec.application.dynamicPluginsConfigMapName` CR field):
```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: dynamic-plugins-rhdh
data:
  dynamic-plugins.yaml: |
    includes:
      - dynamic-plugins.default.yaml
    plugins:
      # [...]
      # Enable Bulk import plugins.
      - package: ./dynamic-plugins/dist/janus-idp-backstage-plugin-bulk-import-backend-dynamic
        disabled: false
      - package: ./dynamic-plugins/dist/janus-idp-backstage-plugin-bulk-import
        disabled: false
```

And (example CR):
```yaml
apiVersion: rhdh.redhat.com/v1alpha1
kind: Backstage
metadata:
  name: my-rhdh
spec:
  application:
    appConfig:
      configMaps:
        - name: app-config-rhdh
    dynamicPluginsConfigMapName: dynamic-plugins-rhdh
  # [...]
```

Author: I am documenting here only the content in the your `dynamic-plugins.yaml` file. The rest should be better documented in the plugins book

Author: ```suggestion
See link:{installing-and-viewing-dynamic-plugins-url}[{installing-and-viewing-dynamic-plugins-title}].
```

[id="enabling-ang-giving-access-to-the-bulk-import-feature"]
= Enabling and giving access to the Bulk Import feature

.Procedure
. To give {product-short} access to GitHub, link:{linkgettingstartedguide}#configuring-github-integration[configure GitHub integration].

. The Bulk Import plugins are installed but disabled by default.
To enable the `./dynamic-plugins/dist/janus-idp-backstage-plugin-bulk-import-backend-dynamic` and `./dynamic-plugins/dist/janus-idp-backstage-plugin-bulk-import` plugins,
edit your `dynamic-plugins.yaml` with following content.
See link:{plugins-configure-book-url}[{plugins-configure-book-title}].
+
.`dynamic-plugins.yaml` fragment
[source,yaml]
----

> global:

Reviewer: IMO, this example is too-specific to the Helm Chart (specifically the `global.dynamic` field). 
I'll suggest removing the `global.dynamic` container from this YAML.
But actually, I just came across this nice doc on how to enable a dynamic plugin using either Helm or the Operator: https://docs.redhat.com/en/documentation/red_hat_developer_hub/1.3/html/installing_and_viewing_dynamic_plugins/index
I think we could link to this doc instead. WDYT?

> global:
>   dynamic:
>     includes:
>       - dynamic-plugins.default.yaml
>     plugins:
>       - package: ./dynamic-plugins/dist/janus-idp-backstage-plugin-bulk-import-backend-dynamic
>         disabled: false
>       - package: ./dynamic-plugins/dist/janus-idp-backstage-plugin-bulk-import
>         disabled: false

Author: ```suggestion
plugins:
  - package: ./dynamic-plugins/dist/janus-idp-backstage-plugin-bulk-import-backend-dynamic
    disabled: false
  - package: ./dynamic-plugins/dist/janus-idp-backstage-plugin-bulk-import
    disabled: false
```

----

. Only {product-short} administrators or users with the `bulk.import` permission can use the Bulk Import feature.
Configure the required `bulk.import` RBAC permission for the users that are not administrators.

> See link:{authorization-book-url}#ref-rbac-permission-policies_title-authorization[Permission policies in Red Hat Developer Hub].

Reviewer: I guess an example of external policy file with the bulk-import permission might be helpful here:
```yaml
kind: ConfigMap
apiVersion: v1
metadata:
  name: rbac-policy
  namespace: rhdh
data:
  rbac-policy.csv: |
    p, role:default/bulk-import, bulk.import, use, allow
    p, role:default/bulk-import, catalog-entity.read, read, allow
    p, role:default/bulk-import, catalog.entity.create, create, allow
    g, user:default/<YOUR_USER>, role:default/bulk-import
```

Author: ```suggestion
See link:{authorization-book-url}#ref-rbac-permission-policies_title-authorization[Permission policies in Red Hat Developer Hub].
+
.`rbac-policy.csv` fragment
[source,csv,subs="+quotes"]
----
p, role:default/bulk-import, bulk.import, use, allow
p, role:default/bulk-import, catalog-entity.read, read, allow
p, role:default/bulk-import, catalog.entity.create, create, allow
g, user:default/__<your_user>__, role:default/bulk-import
----
```

> See link:{authorization-book-url}#ref-rbac-permission-policies_title-authorization[Permission policies in Red Hat Developer Hub].

Reviewer: I realize that https://docs.redhat.com/en/documentation/red_hat_developer_hub/1.3/html-single/authorization/index#ref-rbac-permission-policies_title-authorization does not list "Bulk Import permissions" as policies supported in Developer Hub. 
I think it should be added to the list as well, similar to the tables for Scaffolder, Topology, OCM, ... permissions.
Here is the upstream table: https://github.com/janus-idp/backstage-plugins/blob/release-1.3/plugins/rbac-backend/docs/permissions.md#bulk-import

Author: Added

[id="enabling-and-giving-access-to-the-bulk-import-feature"]
= Enabling and giving access to the Bulk Import feature

> You can enable the Bulk Import feature for users and give them the necessary permissions to access it.

Reviewer 2: Please add some spaces here, as it's not rendering correctly.

![image](https://github.com/user-attachments/assets/d3e8ee0e-223f-43b2-9bc2-4642bd708073)

Author: ```suggestion
You can enable the Bulk Import feature for users and give them the necessary permissions to access it.

```

.Prerequisites
* You have link:{authentication-book-url}#enabling-authentication-with-github[configured GitHub authentication and integration].

.Procedure

. The Bulk Import plugins are installed but disabled by default.
To enable the `./dynamic-plugins/dist/janus-idp-backstage-plugin-bulk-import-backend-dynamic` and `./dynamic-plugins/dist/janus-idp-backstage-plugin-bulk-import` plugins,
edit your `dynamic-plugins.yaml` with the following content:
+
.`dynamic-plugins.yaml` fragment
[source,yaml]
----
plugins:
  - package: ./dynamic-plugins/dist/janus-idp-backstage-plugin-bulk-import-backend-dynamic
    disabled: false
  - package: ./dynamic-plugins/dist/janus-idp-backstage-plugin-bulk-import
    disabled: false
----
+
See link:{installing-and-viewing-dynamic-plugins-url}[{installing-and-viewing-dynamic-plugins-title}].

. Configure the required `bulk.import` RBAC permission for the users who are not administrators as follows:
+
.`rbac-policy.csv` fragment
[source,csv,subs="+quotes"]
----
p, role:default/bulk-import, bulk.import, use, allow
p, role:default/bulk-import, catalog-entity.read, read, allow

> p, role:default/bulk-import, catalog.entity.create, create, allow

Reviewer 3: `catalog-entity.read` and `catalog.entity.create` are not required as far I know for bulk import. cc @rm3l @debsmita1

Reviewer: This was meant to be a full example I copied from my own local config. But you're right, it might be confusing. Let's remove the `p, role:default/bulk-import, catalog-entity.read, read, allow` and `p, role:default/bulk-import, catalog.entity.create, create, allow` lines.

> p, role:default/bulk-import, catalog-entity.read, read, allow
> p, role:default/bulk-import, catalog.entity.create, create, allow

Author: ```suggestion
```

> [id="enabling-ang-giving-access-to-the-bulk-import-feature"]

Author: ```suggestion
[id="enabling-and-giving-access-to-the-bulk-import-feature"]
```

= Enabling and giving access to the Bulk Import feature

> 

Reviewer 2: ```suggestion
You can enable the Bulk Import feature for users and give them the necessary permissions to access it.
```

.Prerequisites

> . You have link:{authentication-book-url}#enabling-authentication-with-github[configured GitHub authentication and integration].

Author: ```suggestion
* You have link:{authentication-book-url}#enabling-authentication-with-github[configured GitHub authentication and integration].
```

> .Procedure
> . To give {product-short} access to GitHub, link:{linkgettingstartedguide}#configuring-github-integration[configure GitHub integration].

Author: ```suggestion
.Prerequisites
. You have link:{authentication-book-url}#enabling-authentication-with-github[configured GitHub authentication and integration].

.Procedure
```

.Procedure

. The Bulk Import plugins are installed but disabled by default.
To enable the `./dynamic-plugins/dist/janus-idp-backstage-plugin-bulk-import-backend-dynamic` and `./dynamic-plugins/dist/janus-idp-backstage-plugin-bulk-import` plugins,

> edit your `dynamic-plugins.yaml` with following content.

Reviewer 2: ```suggestion
edit your `dynamic-plugins.yaml` with the following content:
```

+
.`dynamic-plugins.yaml` fragment
[source,yaml]
----
plugins:
  - package: ./dynamic-plugins/dist/janus-idp-backstage-plugin-bulk-import-backend-dynamic
    disabled: false
  - package: ./dynamic-plugins/dist/janus-idp-backstage-plugin-bulk-import
    disabled: false
----
+
See link:{installing-and-viewing-dynamic-plugins-url}[{installing-and-viewing-dynamic-plugins-title}].

> . Only {product-short} administrators or users with the `bulk.import` permission can use the Bulk Import feature.

Reviewer 2: ```suggestion
. Configure the required `bulk.import` RBAC permission for the users who are not administrators as follows:
```

> Configure the required `bulk.import` RBAC permission for the users that are not administrators.

Reviewer 2: ```suggestion

```

> See link:{authorization-book-url}#ref-rbac-permission-policies_title-authorization[Permission policies in Red Hat Developer Hub].

Author: ```suggestion
```

----
p, role:default/bulk-import, bulk.import, use, allow
p, role:default/bulk-import, catalog-entity.read, read, allow
p, role:default/bulk-import, catalog.entity.create, create, allow
g, user:default/__<your_user>__, role:default/bulk-import
----

> 

Author: ```suggestion
+
See link:{authorization-book-url}#ref-rbac-permission-policies_title-authorization[Permission policies in Red Hat Developer Hub].
```

g, user:default/__<your_user>__, role:default/bulk-import
----
+

> See link:{authorization-book-url}#ref-rbac-permission-policies_title-authorization[Permission policies in Red Hat Developer Hub].

Reviewer 2: ```suggestion
Note that only {product-short} administrators or users with the `bulk.import` permission can use the Bulk Import feature. See link:{authorization-book-url}#ref-rbac-permission-policies_title-authorization[Permission policies in Red Hat Developer Hub].
```

[id="enabling-ang-giving-access-to-the-bulk-import-feature"]
= Enabling and giving access to the Bulk Import feature

.Prerequisites
. You have link:{authentication-book-url}#enabling-authentication-with-github[configured GitHub authentication and integration].

.Procedure

. The Bulk Import plugins are installed but disabled by default.
To enable the `./dynamic-plugins/dist/janus-idp-backstage-plugin-bulk-import-backend-dynamic` and `./dynamic-plugins/dist/janus-idp-backstage-plugin-bulk-import` plugins,
edit your `dynamic-plugins.yaml` with following content.
+
.`dynamic-plugins.yaml` fragment
[source,yaml]
----
plugins:
  - package: ./dynamic-plugins/dist/janus-idp-backstage-plugin-bulk-import-backend-dynamic
    disabled: false
  - package: ./dynamic-plugins/dist/janus-idp-backstage-plugin-bulk-import
    disabled: false
----

> See link:{installing-and-viewing-dynamic-plugins-url}[{installing-and-viewing-dynamic-plugins-title}].

Author: ```suggestion
+
See link:{installing-and-viewing-dynamic-plugins-url}[{installing-and-viewing-dynamic-plugins-title}].

```

+
.`dynamic-plugins.yaml` fragment
[source,yaml]
----
plugins:
  - package: ./dynamic-plugins/dist/janus-idp-backstage-plugin-bulk-import-backend-dynamic
    disabled: false
  - package: ./dynamic-plugins/dist/janus-idp-backstage-plugin-bulk-import
    disabled: false
----
+
See link:{installing-and-viewing-dynamic-plugins-url}[{installing-and-viewing-dynamic-plugins-title}].

. Only {product-short} administrators or users with the `bulk.import` permission can use the Bulk Import feature.
Configure the required `bulk.import` RBAC permission for the users that are not administrators.
+
.`rbac-policy.csv` fragment
[source,csv,subs="+quotes"]
----
p, role:default/bulk-import, bulk.import, use, allow
p, role:default/bulk-import, catalog-entity.read, read, allow
p, role:default/bulk-import, catalog.entity.create, create, allow
g, user:default/__<your_user>__, role:default/bulk-import
----
+
See link:{authorization-book-url}#ref-rbac-permission-policies_title-authorization[Permission policies in Red Hat Developer Hub].

.Verification
* The sidebar displays a *Bulk Import* line.
* The *Bulk Import* page shows a list of *Added Repositories*.

> .Troubleshooting
> * If clicking on the *Bulk Import* option displays the following error message: 
> _To view the added repositories, contact your administrator to give you the `bulk.import` permission_, which means the user does not have the necessary permission to use the Bulk Import feature.
> See link:{authorization-book-url}#ref-rbac-permission-policies_title-authorization[Permission policies in Red Hat Developer Hub] for more information on how an administrator can add that permission to your user.

Author: Removing. It repeats the last procedure step, and the error message is self-sufficient.

```suggestion
```
